RecoEgamma/EgammaTools/plugins/CalibratedElectronProducers.cc

```diff
@@ -0,0 +1,198 @@
+  recHitCollectionEEToken_(consumes<EcalRecHitCollection>(conf.getParameter<edm::InputTag>("recHitCollectionEE"))),
+  produceCalibratedObjs_(conf.getParameter<bool>("produceCalibratedObjs"))
+{
+  energyCorrector_.setMinEt(conf.getParameter<double>("minEtToCalibrate"));  
+  
+  if (conf.getParameter<bool>("semiDeterministic")) {
+     semiDeterministicRng_.reset(new TRandom2());
+     energyCorrector_.initPrivateRng(semiDeterministicRng_.get());
+  }
+
+  if(produceCalibratedObjs_) produces<objectVector>();
+  
+  for(const auto& toStore : valMapsToStore_){
+    produces<floatMap>(toStore.second);
+  }
+}
+
+template<typename T>
+void
+CalibratedElectronProducerT<T>::produce( edm::Event & iEvent, const edm::EventSetup & iSetup ) 
+{
+  
+  epCombinationTool_.setEventContent(iSetup);
+  
+  edm::Handle<objectCollection> inHandle;
+  iEvent.getByToken(electronToken_, inHandle);
+  
+  edm::Handle<EcalRecHitCollection> recHitCollectionEBHandle;
+  edm::Handle<EcalRecHitCollection> recHitCollectionEEHandle;
+  
+  iEvent.getByToken(recHitCollectionEBToken_, recHitCollectionEBHandle);
+  iEvent.getByToken(recHitCollectionEEToken_, recHitCollectionEEHandle);
+  
+  std::unique_ptr<objectVector> out = std::make_unique<objectVector>();
+
+  size_t nrObj = inHandle->size();
+  std::vector<std::vector<float> > results(EGEnergySysIndex::kNrSysErrs);
+  for(auto& res : results) res.reserve(nrObj);
+
+  const ElectronEnergyCalibrator::EventType evtType = iEvent.isRealData() ? 
+    ElectronEnergyCalibrator::EventType::DATA : ElectronEnergyCalibrator::EventType::MC; 
+  
+  
+  for (auto &ele : *inHandle) {
+    out->push_back(ele);
+    
+    if(semiDeterministicRng_) setSemiDetRandomSeed(iEvent,ele,nrObj,out->size());
+
+    const EcalRecHitCollection* recHits = (ele.isEB()) ? recHitCollectionEBHandle.product() : recHitCollectionEEHandle.product();
+    std::vector<float> uncertainties = energyCorrector_.calibrate(out->back(), iEvent.id().run(), recHits, iEvent.streamID(), evtType);
+    
+    for(size_t index=0;index<EGEnergySysIndex::kNrSysErrs;index++){
+      results[index].push_back(uncertainties[index]);
+    }
+  }
+  if(produceCalibratedObjs_){
+    edm::OrphanHandle<objectVector> outHandle = iEvent.put(std::move(out));
+    for(const auto& mapToStore : valMapsToStore_){
+      fillAndStoreValueMap(iEvent,outHandle,results[mapToStore.first],mapToStore.second);
+    }
+  }else{
+    for(const auto& mapToStore : valMapsToStore_){
+      fillAndStoreValueMap(iEvent,inHandle,results[mapToStore.first],mapToStore.second);
+    }
+  }
+}
+
+template<typename T>
+void CalibratedElectronProducerT<T>::setSemiDetRandomSeed(const edm::Event& iEvent,const T& obj,size_t nrObjs,size_t objNr)
+{
+  if(obj.superCluster().isNonnull()){
+    semiDeterministicRng_->SetSeed(egamma::getRandomSeedFromSC(iEvent,obj.superCluster()));
+  }else{
+    semiDeterministicRng_->SetSeed(egamma::getRandomSeedFromObj(iEvent,obj,nrObjs,objNr));
```

Comment:
So I had thought about this when designing it and I decided it was a desired feature. My reasoning, when I change the smearing, when comparing the changes, I dont want to also include the effects of different random numbers.  So if I change smearing X -> Y, I can actually see the exact effect of the change. So far I havent been able to see any downside to this although that doesnt mean there isnt one. 

I do sympathise with the idea of an additional configurable seed "just in case". I wouldnt want to have a hash of the module name as I want this number reproducible outside CMSSW. Also it would enforce a constaint on the module name. 

I would like to always ensure the electrons and photons are synced random nr wise though. So adding an additional parameter means one more parameter that can diverge. Will have a think about it. 

Right now if you really dont want this effect, you can use a true random number but then of course that breaks ele/pho synchronization.